### PR TITLE
Add service API handler stubs for status, stop and restart

### LINKED ISSUE #3393: api/mgmt: Implement new service management API's (Shutdown, Restart)
## Expected Behavior
Management API to cater for `minio control service` command which was previously removed.

## Current Behavior
No way to restart or shutdown a running cluster of servers.

## Possible Solution

A new API perhaps like this. 

 - Service(type serviceType) 

```
type serviceType enum {
      Shutdown
      Restart
      Status
}
```

## Context
To bring in management functionalities through this API.

### PR DESCRIPTION
**Pull request by krisis**
Add service REST API handlers for status, stop and restart operations. These operations are apply uniformly to single node FS, XL and distributed setup. In a distributed setup, these operations are applied to all the nodes participating in the cluster. i.e a stop command would stop all the minio servers.

## Description

Adds service REST API handling. In a distributed setup, as part of handling the API, other nodes are intimated of the operation via RPC.

## Motivation and Context

These APIs will be consumed by ``mc admin service`` subcommands to manage a minio setup.

## How Has This Been Tested?

A sample go program which makes the corresponding REST calls was used to test, besides unit tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Fixes #3393

### CODE REVIEW

**Comment by krisis** [comment]
This PR still has an issue. A Stop/Restart service REST API call would stop/restart the node to which the request was sent. There is an issue with internal admin RPC registration. I am working on this. Meanwhile reviews can start.

**Comment by krisis** [comment]
The admin RPC issue is fixed. It was an issue with the order in which routes where registered on the mux (gorilla/mux).

File: cmd/admin-rpc-client.go [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+	args := GenericArgs{}
+	reply := GenericReply{}
+	err := rc.Call("Service.Restart", &args, &reply)
+	if err != nil && err == rpc.ErrShutdown {
+		rc.Close()
+	}
+	return err
+}
+
+type controlPeer struct {
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go]: we can avoid the use of "control" word here..

File: cmd/admin-rpc-client.go [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	addr    string
+	svcClnt stopRestarter
+}
+type controlPeers []controlPeer
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go]: We moved to calling these admin so we should name them as such.

File: cmd/service.go [hunk 3]
```diff
@@ -80,36 +80,37 @@ func (m *ServerMux) handleServiceSignals() error {
 		globalServiceDoneCh <- struct{}{}
 	}
 
-	// Start listening on service signal. Monitor signals.
+	// Wait for SIGTERM in a go-routine.
 	trapCh := signalTrap(os.Interrupt, syscall.SIGTERM)
+	go func(<-chan bool) {
```
Comment (harshavardhana) [inline cmd/service.go:85]: is there a particular reason to remove select usage?

File: cmd/admin-handlers_test.go [hunk 4]
```diff
@@ -68,7 +68,7 @@ func (c cmdType) toServiceSignal() serviceSignal {
 	return serviceStatus
 }
 
-func serviceSignalReceiver(cmd cmdType, t *testing.T) {
+func mockServiceSignalReceiver(cmd cmdType, t *testing.T) {
```
Comment (harshavardhana) [inline cmd/admin-handlers_test.go]: we can avoid the name of mock and use "initTestServiceSignalReceiver"

File: cmd/admin-handlers.go [hunk 5]
```diff
@@ -16,42 +16,41 @@
...
 )
 
 func (adminAPI adminAPIHandlers) ServiceStatusHandler(w http.ResponseWriter, r *http.Request) {
-	fmt.Println("status: ", r)
 	adminErr := checkRequestAuthType(r, "", "", "")
 	if adminErr != ErrNone {
 		writeErrorResponse(w, r, adminErr, r.URL.Path)
 		return
 	}
+	// FIXME: Servicestatus command is not yet implemented in handleServiceSignals
```
Comment (harshavardhana) [inline cmd/admin-handlers.go]: Is this implemented?

File: cmd/admin-handlers.go [hunk 6]
```diff
@@ -26,7 +26,7 @@ const (
 	minioAdminOpHeader   = "X-Minio-Operation"
 )
 
-func (api objectAPIHandlers) ServiceStatusHandler(w http.ResponseWriter, r *http.Request) {
+func (adminAPI adminAPIHandlers) ServiceStatusHandler(w http.ResponseWriter, r *http.Request) {
 	fmt.Println("status: ", r)
 	adminErr := checkRequestAuthType(r, "", "", "")
```
Comment (harshavardhana) [inline cmd/admin-handlers.go]: adminAPIErr

File: cmd/admin-rpc-server.go [hunk 7]
```diff
@@ -16,12 +16,20 @@
 
 package cmd
 
+import (
+	"net/rpc"
+
+	router "github.com/gorilla/mux"
+)
+
+const servicePath = "/service"
```
Comment (harshavardhana) [inline cmd/admin-rpc-server.go]: We can move this to "/admin/service"

**Review by harshavardhana — COMMENTED** [review]
Initial comments.

File: cmd/service.go (hunk 3 above)
Comment (krisis) [reply cmd/service.go:85]: The way we were handling os signals required us to have globalServiceSignalCh buffered. Separating in this fashion allows us to use an unbuffered channel. Unbuffered channels are easy to reason with and write deterministic unit tests.

File: cmd/admin-rpc-client.go (hunk 1 above)
Comment (krisis) [reply cmd/admin-rpc-client.go]: Yes, I shall change it.

File: cmd/admin-rpc-client.go (hunk 2 above)
Comment (krisis) [reply cmd/admin-rpc-client.go]: OK

File: cmd/admin-rpc-server.go (hunk 7 above)
Comment (krisis) [reply cmd/admin-rpc-server.go]: OK.

File: cmd/admin-handlers.go (hunk 6 above)
Comment (krisis) [reply cmd/admin-handlers.go]: done.

File: cmd/admin-handlers.go (hunk 5 above)
Comment (krisis) [reply cmd/admin-handlers.go]: This is an old comment. It is implemented and the FIXME is removed.

File: cmd/admin-handlers_test.go (hunk 4 above)
Comment (krisis) [reply cmd/admin-handlers_test.go]: It is mocking the signal receiver code in service.go and named accordingly. It is not initializing anything so it doesn't make sense naming it so.

File: docs/service-api.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# Service REST API
```
Comment (harshavardhana) [inline docs/service-api.md]: Move this into `docs/admin-api/service.md` ?

File: cmd/admin-rpc-client.go [hunk 9]
```diff
@@ -117,7 +117,7 @@ func initGlobalAdminPeers(eps []*url.URL) {
 	globalAdminPeers = makeControlPeers(eps)
 }
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go:135]: Let's write comment for each of these functions..

File: cmd/admin-rpc-client.go [hunk 10]
```diff
@@ -65,20 +65,20 @@ func (rc remoteAdminClient) Restart() error {
 	return err
 }
 
-type controlPeer struct {
+type adminPeer struct {
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go:80]: We should write some comments about the actual usage of these peers and the structure used here.

File: cmd/admin-handlers_test.go [hunk 11]
```diff
@@ -96,6 +96,14 @@ func testServicesCmdHandler(cmd cmdType, t *testing.T) {
 		t.Fatalf("Unable to initialize server config. %s", err)
 	}
 	defer removeAll(rootPath)
+
+	// Initialize admin peers to make admin RPC calls.
+	eps, err := parseStorageEndpoints([]string{"http://localhost"})
```
Comment (harshavardhana) [inline cmd/admin-handlers_test.go:101]: what will be port used here? "9000"?

File: cmd/admin-rpc-client.go [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+package cmd
+
+import (
+	"net/rpc"
+	"net/url"
+	"path"
+	"sync"
+)
+
+type localAdminClient struct {
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go:27]: Approach looks good,  add comments..

File: cmd/admin-rpc-client.go [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+}
+
+type controlPeer struct {
+	addr    string
+	svcClnt stopRestarter
+}
+type controlPeers []controlPeer
+
+func makeControlPeers(eps []*url.URL) controlPeers {
+	var ret []controlPeer
```
Comment (harshavardhana) [inline cmd/admin-rpc-client.go]: what is "ret" ? we could use a non generic name..

File: cmd/admin-handlers_test.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/admin-handlers_test.go]: My concern is the name itself, let's not use `mock`..  We can prefix it with  `test`  rather than `mock`.

**Review by harshavardhana — COMMENTED** [review]
Few more..

File: docs/service-api.md (hunk 8 above)
Comment (krisis) [reply docs/service-api.md]: OK.

File: cmd/admin-handlers_test.go (hunk 11 above)
Comment (krisis) [reply cmd/admin-handlers_test.go:101]: parseStorageEndpoints() defaults to 9000 if no port is supplied.

**Pull request merged**